chrome/content/zotero/xpcom/translation/translators.js

```diff
@@ -267,76 +270,92 @@ Zotero.Translators = new function() {
 	/**
 	 * Gets web translators for a specific location
 	 * @param {String} uri The URI for which to look for translators
+	 * @param {String} rootUri The root URI of the page, different from `uri` if running in an iframe
 	 */
-	this.getWebTranslatorsForLocation = function(uri) {
-		return this.getAllForType("web").then(function(allTranslators) {
+	this.getWebTranslatorsForLocation = function(URI, rootURI) {
+		var isFrame = URI !== rootURI;
+		var type = isFrame ? "webWithTargetAll" : "web";
+		
+		return this.getAllForType(type).then(function(allTranslators) {
 			var potentialTranslators = [];
+			var translatorConverterFunctions = [];
 			
-			var properHosts = [];
-			var proxyHosts = [];
+			var rootSearchURIs = this.getSearchURIs(rootURI);
+			var frameSearchURIs = isFrame ? this.getSearchURIs(URI) : rootSearchURIs;
 			
-			var properURI = Zotero.Proxies.proxyToProper(uri);
-			var knownProxy = properURI !== uri;
-			if(knownProxy) {
-				// if we know this proxy, just use the proper URI for detection
-				var searchURIs = [properURI];
-			} else {
-				var searchURIs = [uri];
-				
-				// if there is a subdomain that is also a TLD, also test against URI with the domain
-				// dropped after the TLD
-				// (i.e., www.nature.com.mutex.gmu.edu => www.nature.com)
-				var m = /^(https?:\/\/)([^\/]+)/i.exec(uri);
-				if(m) {
-					// First, drop the 0- if it exists (this is an III invention)
-					var host = m[2];
-					if(host.substr(0, 2) === "0-") host = host.substr(2);
-					var hostnames = host.split(".");
-					for(var i=1; i<hostnames.length-2; i++) {
-						if(TLDS[hostnames[i].toLowerCase()]) {
-							var properHost = hostnames.slice(0, i+1).join(".");
-							searchURIs.push(m[1]+properHost+uri.substr(m[0].length));
-							properHosts.push(properHost);
-							proxyHosts.push(hostnames.slice(i+1).join("."));
-						}
-					}
-				}
-			}
+			Zotero.debug("Translators: Looking for translators for "+Object.keys(frameSearchURIs).join(', '));
 			
-			Zotero.debug("Translators: Looking for translators for "+searchURIs.join(", "));
-			
-			var converterFunctions = [];
-			for(var i=0; i<allTranslators.length; i++) {
-				for(var j=0; j<searchURIs.length; j++) {
-					if((!allTranslators[i].webRegexp
-							&& allTranslators[i].runMode === Zotero.Translator.RUN_MODE_IN_BROWSER)
-							|| (uri.length < 8192 && allTranslators[i].webRegexp.test(searchURIs[j]))) {
-						// add translator to list
-						potentialTranslators.push(allTranslators[i]);
-						
-						if(j === 0) {
-							if(knownProxy) {
-								converterFunctions.push(Zotero.Proxies.properToProxy);
-							} else {
-								converterFunctions.push(null);
+			for (let translator of allTranslators) {
+				translatorLoop:
+				for (let rootSearchURI in rootSearchURIs) {
+					let isGeneric = (!translator.webRegexp.root && translator.runMode === Zotero.Translator.RUN_MODE_IN_BROWSER);
```

Comment:
I don't think `RUN_MODE_IN_BROWSER` is related to generic, so I don't think it should be combined (which also makes the below condition rather confusing). Can use a different variable for the other checks.